proxy/proxy.go

```diff
@@ -703,14 +729,39 @@ func (p *Proxy) do(ctx *context) error {
 		}
 
 		backendStart := time.Now()
-		rsp, err := p.makeBackendRequest(ctx)
-		if err != nil {
+		rsp, perr := p.makeBackendRequest(ctx)
+		if perr != nil {
 			if done != nil {
 				done(false)
 			}
 
 			p.metrics.IncErrorsBackend(ctx.route.Id)
-			return err
+
+			neterr := perr.NetError()
+			if neterr != nil && !neterr.Temporary() && ctx.route.IsLoadBalanced {
+				// here we do a transparent retry, because we know it's safe to do
```

Comment:
I would rather be safe than sorry and explicitly list the conditions where we can safely retry. Going to `syscall_unix.go` just to find out which errors are considered temporary isn't really maintainable in my opinion.

Additionally, I'm not sure how is the connection timeout case covered by this. We *should* retry on connection timeouts, since this is what was happening in the original issue, but we *mustn't* retry on read/write timeouts after the connection was established, since it might be dangerous if we retry a non-idempotent request. However, if I'm reading this correctly, both of these would be covered by `neterr.Timeout()` which marks the error as `Temporary()`.

Reply:
No the problem was "connection refused", which is not a neterr.Timeout() but a "not neterr.Temporary()"

Reply:
While an instance is rebooting, you wouldn't get connection resets, since there's nobody there to send a packet back, so you'll get connection timeout. You should be able to test this by blocking incoming packets with `-j DROP`, I think.

My actual concern is that since we retry all requests here, even `POST`, `PUT` and `DELETE` ones, we have to be very careful as to when we retry. This code makes the retry logic completely unclear and dependent on stuff that isn't really documented unless you go into Go source code, without any guarantees on whether it'll change or not. I think that the retry criteria should be made much more clear and documented.

Reply:
I share your concern, but the retry is only happening on permanent network errors, so this should be safe in any case. A retry strategy which is more fine grained have to be per route, because only the user knows if HTTP GET is safe to reapply for temporary errors or timeouts.
I agree that iptables -j DROP is also a valuable case, but in the VM based environment a EC2 instance can die and the hyp can return the connection refused. In any case I want to start with connection refused and later think about other cases we might want to retry. Feel free to write an issue in this repository to specify which else condition(s) you think we should retry.
Thanks for your feedback.

Reply:
I didn't want to suggest making the retry strategy fine grained, sorry if I wasn't clear enough. What I was trying to say is that since the retry strategy applies to all requests, not even those that should be safe enough to retry (and people also do crazy things with GET!), we should be as explicit as possible in the logic to avoid accidentally retrying something. The retry logic should be clear and understandable from the first glance, and shouldn't change whenever the Golang stdlib changes something. The current code which implicitly trusts the `NetError` implementation to provide the correct information isn't really suitable, especially considering that the error might be raised long after the connection was successful and we've sent the request.

Reply:
Also, the hypervisor won't return connection refused if the VM is down — this is only something the VM kernel can do. You can verify it yourself by rebooting or terminating an instance and trying to connect.

Reply:
I honestly think that something you have to trust and what would you suggest to fix it without a trust?

I see these possibilities, maybe you have opinions on these or have to add another one:
1. check error messages to filter for tcp handshake errors (for example "connection refused" or some connection timeout) or tls handshake errors and 
1. have a low timeout for tcp/tls connections, have high timeouts for all others, measure time and hope that if we have passed the tcp/tls timeouts it is safe to retry
1. create the connections pools ourselves
1. do not retry at all

The "check error messages" idea could be dropped as soon Go would have exported error types that help to clearly differentiate, but right now it would required to check strings in errors, which I consider a bit ugly but I would do it if you think it makes more sense.